### DISCUSSION TITLE
Porting the library to STM32CubeIDE

### DISCUSSION THREAD

**Discussion by HamzaHajeir** [jgromes/RadioLib#300] [General]
Hi

Thank you for this kind and great library @jgromes.

I'm using SX1278/76 in a product, I'm using this library flashing ESP8266. And it works great!
However, the other node is programmed over STM32CubeIDE, using HAL libraries.
I'm using a workable library, but it doesn't support asynchronous receiving/transmitting of messages as RadioLib does.

In my to-do's, I'm improving and enhancing a protocol, for the protocol to be developed, as it requires async receiving of messages, So I've decided to port this library (SX127x module) to STM32 HAL. 

What should I consider for this work to be done? 

And I'm aware that I'd need to port the interrupt handlers, what else needed to be ported?

Any comment, guidance, and suggestion is welcomed!

**Comment by jgromes** [comment]
Essentially, you'd have to port everything from Arduino that is used in Module.cpp, which is:

* SPIClass and SPISettings
* GPIO functions like pinMode or digitalWrite
* timing functions like micros, millis, delay etc.

So you'd have to make your own implementation of Arduino API. The question then arises - can't you use [stm32duino](https://github.com/stm32duino/Arduino_Core_STM32)? That project also uses STM32 HAL, but builds an Arduino layer on top of it, which is pretty much what you would be doing.

**Reply by HamzaHajeir** [reply]
I'll need to investigate SPI class stuff, others (GPIO and timing) are easier to me. 
What's considered harder is Interrupt.

I've optimized my low power application on top of HAL, which pays its price well in turn.

**Reply by jgromes** [reply]
You can still use all the HAL stuff with stm32duino, you just wouldn't have to write all of the Arduino implementation yourself.

**Reply by HamzaHajeir** [reply]
Of-course, but I've 2 issues: 
1. it's not recommended from developers (I don't REALLY know why with proofs), but using Arduino would increase binary code significantly, beside "unnecessary" other executed codes that aimed generalization.

2. My application is already optimized. 

I think trying to "steal" the necessary Arduino files related to SPI, IO, timing are the necessary ones. Currently I don't know where it would stop.

**Reply by HamzaHajeir** [reply]
Because having the necessary Arduino APIs on top of STM32CubeIDE goes unmanaged, I've decided to port this library itself, deleting everything unrelated to SX1278.

However, I've noticed this little 'bug' :) : 
https://github.com/jgromes/RadioLib/blob/e27c3ddef525a913ce815ba32e908d4cda90757a/src/Module.h#L330

* That's except if it's used in alternate with _rx. per Module

**Reply by jgromes** [reply]
It's not a bug, _rx is used for an extra GPIO pin on SPI modules (since those won't have a UART interface).

>  I've decided to port this library itself, deleting everything unrelated to SX1278.

I would strongly advise against that, since you will lose the ability to easily update the library in case new patches are added. But go ahead if that seems the most reasonable to you, it's not like I can stop you ;)

**Reply by HamzaHajeir** [reply]
I see, the issue is porting to STM32CubeIDE without using stm32duino is nearly impossible.

I'll think better, I've just got a thought is that I'd only need an async receive (async transmit is a plus), so a tweak to that library could solve it.
